examples/hmm_enum.py

```diff
@@ -0,0 +1,279 @@
+joint density.
+
+To marginalize out discrete variables ``x``:
+
+1. Verify that the variable dependency structure in your model
+    admits tractable inference, i.e. the dependency graph among
+    enumerated variables should have narrow treewidth.
+2. Ensure your model can handle broadcasting of the sample values
+    of those variables
+
+Note that difference from [1], which uses Python loop, here we use
+:func:`~numpryo.contrib.control_flow.scan` to reduce compilation
+times (only one step needs to be compiled) of the model. Under the
+hood, `scan` stacks all the priors' parameters and values into
+an additional time dimension. This allows us computing the joint
+density in parallel. In addition, the stacked form allows us
+to use the parallel-scan algorithm in [2], which reduces parallel
+complexity from O(length) to O(log(length)).
+
+**References:**
+
+    1. *Pyro's Hidden Markov Model example*,
+       (https://pyro.ai/examples/hmm.html)
+    2. *Temporal Parallelization of Bayesian Smoothers*,
+       Simo Sarkka, Angel F. Garcia-Fernandez
+       (https://arxiv.org/abs/1905.13002)
+"""
+
+import argparse
+import logging
+import pickle
+import time
+
+from jax import random
+import jax.numpy as jnp
+
+import numpyro
+from numpyro.contrib.control_flow import scan
+from numpyro.contrib.indexing import Vindex
+import numpyro.distributions as dist
+from numpyro.handlers import mask
+from numpyro.infer import HMC, MCMC, NUTS
+
+
+# Let's start with a simple Hidden Markov Model.
+#
+#     x[t-1] --> x[t] --> x[t+1]
+#        |        |         |
+#        V        V         V
+#     y[t-1]     y[t]     y[t+1]
+#
+# This model includes a plate for the data_dim = 44 keys on the piano. This
+# model has two "style" parameters probs_x and probs_y that we'll draw from a
+# prior. The latent state is x, and the observed state is y.
+def model_1(sequences, lengths, args, include_prior=True):
+    num_sequences, max_length, data_dim = sequences.shape
+    with mask(mask_array=include_prior):
+        probs_x = numpyro.sample("probs_x",
+                                 dist.Dirichlet(0.9 * jnp.eye(args.hidden_dim) + 0.1)
+                                     .to_event(1))
+        probs_y = numpyro.sample("probs_y",
+                                 dist.Beta(0.1, 0.9)
+                                     .expand([args.hidden_dim, data_dim])
+                                     .to_event(2))
+
+    def transition_fn(carry, y):
+        x_prev, t = carry
+        with numpyro.plate("sequences", num_sequences, dim=-2):
+            with mask(mask_array=(t < lengths)[..., None]):
+                x = numpyro.sample("x", dist.Categorical(probs_x[x_prev]))
+                with numpyro.plate("tones", data_dim, dim=-1):
+                    numpyro.sample("y", dist.Bernoulli(probs_y[x.squeeze(-1)]), obs=y)
+        return (x, t + 1), None
+
+    x_init = jnp.zeros((num_sequences, 1), dtype=jnp.int32)
```

Comment:
Instead of `x_init = 0` in Pyro, we need to make `x_init` have the same shape as the output (in non-enumerate version).